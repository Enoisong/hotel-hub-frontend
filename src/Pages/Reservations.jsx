import React, { useState, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { fetchReservationDetailsById, fetchReservations } from '../redux/features/reservations/reservationsSlice'; 

const Reservation = () => {
  const dispatch = useDispatch();
  const user = useSelector((state) => state.user); 
  const reservations = useSelector((state) => state.reservations);  
  const rooms = useSelector((state) => state.rooms);  

  const [name, setName] = useState('');
  const [date, setDate] = useState('');
  const [selectedRoomId, setSelectedRoomId] = useState('');

  useEffect(() => {
    // Fetch user's reservations when the component mounts
    dispatch(fetchReservations(user.id));
    // Fetch available rooms when the component mounts
    dispatch(fetchRooms());
  }, [dispatch, user.id]);

  useEffect(() => {
    // Autofill name when a room is selected
    const selectedRoom = rooms.find(room => room.id === selectedRoomId);
    if (selectedRoom) {
      setName(selectedRoom.name);
    }
  }, [selectedRoomId, rooms]);

  const handleReservationSubmit = () => {
    // Create a new reservation
    dispatch(createReservation({ name, date, roomId: selectedRoomId, userId: user.id }));
  };
  return (
    <div>
      <h1>Booking</h1>
      <div className="reservation-form">
        <label>Name:</label>
        <input
          type="text"
          value={name}
          onChange={(e) => setName(e.target.value)}
        />
        <label>Date:</label>
        <input
          type="date"
          value={date}
          onChange={(e) => setDate(e.target.value)}
        />
        <label>Select a Room:</label>
        <select
          value={selectedRoomId}
          onChange={(e) => setSelectedRoomId(e.target.value)}
        >
          <option value="">Select a room</option>
          {rooms.map((room) => (
            <option key={room.id} value={room.id}>
              {room.name}
            </option>
          ))}
        </select>
        <button onClick={handleReservationSubmit}>Reserve</button>
      </div>

      <h2>My Reservations</h2>
      <ul>
        {reservations.map((reservation) => (
          <li key={reservation.id}>
            <p>Name: {reservation.name}</p>
            <p>Date: {reservation.date}</p>
          </li>
        ))}
      </ul>
    </div>
  );
};

export default Reservation;
