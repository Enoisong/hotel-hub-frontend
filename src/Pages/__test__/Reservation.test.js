import React from 'react';
import { render } from '@testing-library/react';
import Reservations from '../Reservations';

jest.mock('../../components/ReservationForm', () => {
  const ReservationForm = () => <div data-testid="reservation-form" />;
  ReservationForm.displayName = 'ReservationForm';
  return ReservationForm;
});

const ReservationsComponent = () => <Reservations />;

describe('Reservations Component', () => {
  test('renders Reservations component', () => {
    const { asFragment } = render(<ReservationsComponent />);
    expect(asFragment()).toMatchSnapshot();
  });
});
