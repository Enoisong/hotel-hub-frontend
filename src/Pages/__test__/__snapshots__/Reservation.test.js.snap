// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reservations Component renders Reservations component 1`] = `
<DocumentFragment>
  <div
    class="md:container w-full reservation_container flex justify-center"
  >
    <div
      class="relative flex flex-col bg-clip-border rounded-xl bg-white text-gray-700 shadow-md mt-6 w-98 p-5"
    >
      <div
        class="p-6 py-6"
      >
        <div
          data-testid="reservation-form"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
